### PR TITLE
Add image slideshow screensaver mode

### PR DESCRIPTION
**Pull request by csonsino**
This change adds an image slideshow screensaver mode with optional
background audio.  The existing video slideshow has been refactored
to include this new mode.

### CODE REVIEW

**Comment by pjft** [comment]
Thanks for putting this together! Haven't read through the code, but tested it on my Pi 3.

Overall, when everything is set up as expected, it works as intended!

A few notes on the overall user experience out of the box, if I may:

- I started by selecting it from the menu, and my first surprise was that it only showed black. I expected it to run a slideshow of the pictures on my gamelists. Only then did I see the command line log saying that a specific folder was empty. However, had I not been running this from SSH, I wouldn't have seen it and I'd have thought there was a bug. May I suggest presenting the setup instructions as a popup after one selects the SlideShow screensaver, as it sets the expectation from the get go of what images you'll actually be using and how to set it up? If you want to have an explicit entry in the slideshow menu as well, with "Setup Instructions" that when clicked will show the instructions, that's also fine.
- A note on the audio path: may I suggest making it .emulationstation/slideshow/audio, so that all related - and necessary files - are inside the slideshow folder? Or even just .emulationstation/slideshow . Just so we don't create multiple separate folders, and "audio" seems to be a somewhat vague folder name that could end up being used for other purposes in the future.
- For the sake of menu organization, it might make sense to create a "screensaver options menu" and then have two separate sub-menus - one being the video one, and the other being the slideshow one. It can be done later, but it's probably best as having an entry for each screensaver type is probably an overkill and makes the menu too big.
- Also, even if you have the option to read images from your pre-defined folder, why not have the option to just render the images from the user's gamelists, similar to the video screensaver? At the moment, users will need to have actual SSH access to the filesystem to set it up, whereby users can get images on the Pi via scrapers straight away, so it could be an easy win.

Just a few thoughts, all for your consideration, and most can be done at a later stage. :)

I noticed you mention that you wanted the settings to be editable by the user - that was going to be one of my comments. I'll look through the code and share some feedback on that in the coming days.

Thank you very much for this!

File: es-app/src/guis/GuiSlideshowScreensaverOptions.cpp [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	addWithLabel("IMAGE FILTER", sss_image_filter);
+
+ 	auto sss_stretch = std::make_shared<SwitchComponent>(mWindow);
+	sss_stretch->setState(Settings::getInstance()->getBool("SlideshowScreenSaverStretch"));
+	addWithLabel("STRETCH IMAGES", sss_stretch);
+	addSaveFunc([sss_stretch] {
+	    Settings::getInstance()->setBool("SlideshowScreenSaverStretch", sss_stretch->getState());
+	  });
+
+	auto sss_bg_audio_file = std::make_shared<TextComponent>(mWindow);
```
Comment (pjft) [inline es-app/src/guis/GuiSlideshowScreensaverOptions.cpp]: For editing the text, it could be worth taking a look at GuiMetadataEd.cpp, in particular how they handled MD_MULTILINE_STRING editing. Hopefully it gives you some inspiration on how it has been done in other ES locations. I don't have experience with that particular UI pattern, but let me know if you struggle with it and happy to help if I can!

**Comment by csonsino** [comment]
Thanks for the comments.  I should be able to address most of that quickly.

This project has been amazing for my custom built cabinet.  Looking forward to making a few more contributions.

**Comment by pjft** [comment]
By all means, and looking forward to having more of your contributions!

Glad that this has been helping out with your custom built cabinet.

Take my comments as just that - comments. From a personal standpoint, I try
to be focused on the end-user experience delivery, as I'm somewhat mindful
of not everyone being as tech savvy as us though that shouldn't prevent
them from benefiting from all the work everyone does.

Also, I've seen someone mention one time that, as an ideal, ES should be
completely usable by someone running it on an arcade cabinet that's not
connected to anything, and doesn't have a keyboard - so the less one needs
to play with the terminal and SSH, the better. I kind of liked that ideal,
so it's a nice perspective to keep in mind.

So yeah, feel free to ignore some of them, or plan for those developments
in the future, but overall looked good. I'll look through the code in the
coming days, if time allows, and/or if others don't.

Thank you!


- Paulo Tavares

**Comment by csonsino** [comment]
The latest commit should resolve the issues in your comments.

By default, the slideshow screensaver will show game list images.

Thanks for the pointer to GuiMetadataEd.cpp, that helped a lot!  The one thing that I cannot figure out is how to enter text in the popup window.  I can move focus into the text box and scroll left and right to see the current value, but I cannot actually enter new text or modify the existing text- the keyboard does not function as I would expect.

File: es-app/src/SystemScreenSaver.cpp [hunk 2]
```diff
@@ -301,8 +301,8 @@ void SystemScreenSaver::pickGameListNode(unsigned long index, const char *nodeNa
 		pugi::xml_document doc;
 		pugi::xml_node root;
 
-		// We only want images from game systems
-		if (!(*it)->isGameSystem())
+		// We only want nodes from game systems that are not collections
+		if (!(*it)->isGameSystem() || (*it)->isCollection())
```
Comment (pjft) [inline es-app/src/SystemScreenSaver.cpp]: I think that's wise, but in that case isn't the logic reversed? Should it not be

!((*it)->isGameSystem() || (*it)->isCollection())

(missing parentheses around both tests)

That being said, it's probably a non-issue, as collections do not (or should not :P ) have a gamelist path.

But it's certainly more readable, well done.

**Comment by pjft** [comment]
Thanks for doing this so quickly.

I built it and tested it, and it's certainly sharper and nicer from the get go! Well done. Also, great work on refactoring the menus. 

I couldn't do an extensive test of the previous use case (audio and images from the folder) but I'm assuming that they're still working as intended.

I'll take a look at the code for the input when I have a chance, I haven't had the change to work with it. 

Best.

**Comment by pjft** [comment]
Huh. 

It seems I also can't edit any text in the Metadata Editor window - so it's not related to your implementation.

I'd suggest trying to reduce the size of the values, though - they're larger than the caption at the moment.

Let me skim through the commit history and see what might have broken it.

**Comment by pjft** [comment]
EDIT: it actually works, just not when you're running it via SSH or terminal as the keyboard events are being registered in a different console.

If you replace your standard ES binary by yours and reboot it, you'll be able to use the keyboard as intended.

It _is_ weird, but I have no way around it.

TL;DR: it's running as intended, there's no problem with the build.

File: es-app/src/SystemScreenSaver.cpp (hunk 2 above)
Comment (csonsino) [inline es-app/src/SystemScreenSaver.cpp]: In the countGameListNodes() function (used in countVideos() and countImages()), we use:

  if (!(*it)->isCollection() && (*it)->isGameSystem())

to make sure that the node is NOT a collection and IS a game system.

In pickGameListNode(), we use:

  if (!(*it)->isGameSystem() || (*it)->isCollection())

BUT we "continue" (skip over the node) if that matches.  So if the node is NOT a game system, or it IS a collection, then skip the node.

I was having a problem where the Slideshow Screensaver was showing the icons from the "retropie" system, so I decided to add the "isGameSystem()" checks.

I also want to make sure that the checks are consistent when counting the media files, and when deciding if the node is the "random" one.  I don't want to ignore collections when counting media, but use them when actually picking the node.  Honestly, I haven't played around with collections at all, so I don't really understand how they work.  I assume that they are a "virtual game set"- a subset of a real system's games...?

Comment (pjft) [inline es-app/src/SystemScreenSaver.cpp]: Sorry, I'm just tired :)

Reading only a snippet of the code, with the comment "we only want nodes that are not collections" but the condition checking whether it _is_ a collection got me to ask about that. I thought the condition was going to check if they were not collections.

Nevermind.

You are indeed using isGameSystem correctly. RetroPie will not be a GameSystem. 

A collection is really just a group of existing games - a subset, organized in a new top level entry (system) - but keeping the games in their original places.

You can easily test it out by turning on the "All Games" and "Last Played" at least on your menu, if you don't want to play with "Favorites".

Thanks - any questions just ask.
